ui/src/app/components/app-router.tsx

```diff
@@ -38,21 +38,36 @@ const NavRedirect = ({path}) => {
   return null;
 };
 
-export const AppRoute = ({path, data = {}, component: Component}): React.ReactElement => {
+export const AppRoute = ({path, data = {}, guards = [], component: Component}): React.ReactElement => {
   const routeParams = useParams();
   const routeHistory = useHistory();
 
-  return <Route exact={true} path={path} >
-    <Component urlParams={routeParams} routeHistory={routeHistory} routeConfig={data}/>
+  return <Route exact={true} path={path} render={
+    () => {
+      const { redirectPath = null } = fp.find(({allowed}) => !allowed(), guards) || {};
+      return redirectPath
+        ? <NavRedirect path={redirectPath}/>
+        : <Component urlParams={routeParams} routeHistory={routeHistory} routeConfig={data}/>;
+    }}>
   </Route>;
 };
 
 export const ProtectedRoutes = (
   {guards, children}: {guards: Guard[], children: any }): React.ReactElement => {
-  const { redirectPath = null } = fp.find(({allowed}) => !allowed(), guards) || {};
 
-  return redirectPath ? <NavRedirect path={redirectPath}/>
-  : <Fragment>{children}</Fragment>;
+  // Pass the guards to the individual routes. Be sure not to overwrite any existing guards
+  const guardedChildren = fp.flow(
+    fp.flatten,
+    fp.toPairs,
+    fp.map(
+      ([key, element]: [string, React.ReactElement]) => {
+        const {guards: elementGuards = []} = element.props;
+        return React.cloneElement(element, {key, guards: [...elementGuards, ...guards]});
```

Comment:
This is the opposite order of guard execution than what I would expect, i.e. was expecting `[...guards, ...elementGuards]`, which should track with increasing specificity.

Reply:
fixed